docs/source/cpp/streaming_execution.rst

```diff
@@ -305,3 +305,601 @@ Datasets may be scanned multiple times; just make multiple scan
+    ABORT_ON_FAILURE(plan->StartProducing());
+
+    // // collect sink_reader into a Table
+    std::shared_ptr<arrow::Table> response_table;
+
+    ARROW_ASSIGN_OR_RAISE(response_table,
+                        arrow::Table::FromRecordBatchReader(sink_reader.get()));
+
+    std::cout << "Results : " << response_table->ToString() << std::endl;
+
+
+``ConsumingSinkNode``
+---------------------
+
+:class:`arrow::compute::ConsumingSinkNode` is a sink node that owns consuming the data and 
+will not finish until the consumption is finished.  Use SinkNode if you are
+transferring the ownership of the data to another system.  
+Use :class:`arrow::compute::ConsumingSinkNode` if the data is being consumed within the exec 
+plan (i.e. the exec plan should not complete until the consumption has completed).
+
+Example::
+
+    // define a Custom SinkNodeConsumer
+    std::atomic<uint32_t> batches_seen{0};
+    arrow::Future<> finish = arrow::Future<>::Make();
+    struct CustomSinkNodeConsumer : public cp::SinkNodeConsumer {
+
+        CustomSinkNodeConsumer(std::atomic<uint32_t> *batches_seen, arrow::Future<>finish): 
+        batches_seen(batches_seen), finish(std::move(finish)) {}
+        // Consumption logic can be written here
+        arrow::Status Consume(cp::ExecBatch batch) override {
+        // data can be consumed in the expected way
+        // transfer to another system or just do some work 
+        // and write to disk
+        (*batches_seen)++;
+        return arrow::Status::OK();
+        }
+
+        arrow::Future<> Finish() override { return finish; }
+
+        std::atomic<uint32_t> *batches_seen;
+        arrow::Future<> finish;
+        
+    };
+    
+    std::shared_ptr<CustomSinkNodeConsumer> consumer =
+            std::make_shared<CustomSinkNodeConsumer>(&batches_seen, finish);
+
+    arrow::compute::ExecNode *consuming_sink;
+
+    ARROW_ASSIGN_OR_RAISE(consuming_sink, MakeExecNode("consuming_sink", plan.get(),
+        {source}, cp::ConsumingSinkNodeOptions(consumer)));
+
+
+``OrderBySinkNode``
+-------------------
+
+This is an extension to the :class:`SinkNode` definition and provides the ability
+to guarantee the ordering of the stream by providing the,
+:class:`arrow::compute::OrderBySinkNodeOptions`. 
+Here the :class:`arrow::compute::SortOptions` are provided to define which columns 
+are used for sorting and under which criterion.
+
+Example::
+
+    arrow::compute::ExecNode *sink;
+
+    ARROW_ASSIGN_OR_RAISE(sink,
+    arrow::compute::MakeExecNode("order_by_sink", plan.get(),
+    {source}, 
+    arrow::compute::OrderBySinkNodeOptions{
+    /*sort_options*/arrow::compute::SortOptions{
+    {	arrow::compute::SortKey{
+    //Column key(s) to order by and how to order by these sort keys.
+    "a",
+    // Sort Order
+    arrow::compute::SortOrder::Descending 
+    }}},&sink_gen}));
+
+
+SelectK-Node
+------------
+
+There is no Select-K-SinkNode available as an entity within the source, but the behavior 
```

Comment:
Since the "node classes" themselves are really an implementation detail we don't need to call out the fact that a node doesn't actually exist or that we are reusing a node.

Reply:
Yes, @lidavidm also mentioned previously. The nodes will be removed and we will only document the 
options and mention them as operations instead of nodes, "scan", "source", etc.